Add authenticated rotation endpoints documentation (#1622)
openbao/openbao: website/content/docs/internals/rotation.mdx

As a review, OpenBao starts in a _sealed_ state. OpenBao is unsealed by providing the unseal keys.
By default, OpenBao uses a technique known as [Shamir's secret sharing algorithm](https://en.wikipedia.org/wiki/Shamir's_Secret_Sharing)
to split the root key into 5 shares, any 3 of which are required to reconstruct the root

> key. The root key is used to protect the encryption key, which is ultimately used to protect

Reviewer: ```suggestion
key. The root key is used to protect the encryption keyring, which is ultimately used to protect
```

data written to the storage backend.

![Key Rotate](/img/openbao-key-rotate.png)

To enable key rotation, we need to support changing the unseal keys, root key, and the
backend encryption key. 

:::warning

Note: Historically these actions were split into two separate operations, `rekey` and 
`rotate`. To further enhance the clarity and address the ambiguity of the terms all

> key rotation processes described by the `rekey` noun were renamed to `rotate` instead. 

Reviewer: ```suggestion
key rotation processes described by the `rekey` verb were renamed to `rotate/<key-name>`
instead. 
```

:::

> The `rotate` operation is used to generate a new root key. When this is being done,

Reviewer: ```suggestion
The `rotate/root` operation is used to generate a new root key. When this is being done,
```

it is possible to change the parameters of the key splitting, so that the number of shares
and the threshold required to unseal can be changed. To perform a rotation a threshold of
the current unseal keys must be provided. This is to prevent a single malicious operator
from performing a rotation and invalidating the existing root key.

Performing a rotation is fairly straightforward. The rotation must be initialized with the
new parameters for the split and threshold. Once initialized, the current unseal keys must
be provided until the threshold is met. Once met, OpenBao will generate the new root key,
perform the splitting, and re-encrypt the encryption key with the new root key. The new
unseal keys are then provided to the operator, and the old unseal keys are no longer usable.

> The `rotate-keyring` operation is used to change the encryption key used to protect data

Reviewer: ```suggestion
The `rotate/keyring` operation is used to change the encryption key used to protect data
```

written to the storage backend. This key is never provided or visible to operators, who
only have unseal keys. This simplifies the rotation, as it does not require the current

> key holders unlike the `rotate` operation. When `rotate-keyring` is triggered, a new

Reviewer: ```suggestion
key holders unlike the `rotate/root` operation. When `rotate/keyring` is triggered, a new
```

encryption key is generated and added to a keyring. All new values written to the storage
backend would be encrypted with the new key. Old values written with previous encryption
keys can still be decrypted since older keys are saved in the keyring. This allows key
rotation to be done online, without an expensive re-encryption process.

> Both the `rotate` and `rotate-keyring` operations can be done online and in a highly

Reviewer: ```suggestion
Both the `rotate/root` and `rotate/keyring` operations can be done online and in a highly
```
